libdnf5/base/goal.cpp

```diff
@@ -596,6 +601,130 @@ GoalProblem Goal::Impl::add_module_specs_to_goal(base::Transaction & transaction
     return ret;
 }
 
+GoalProblem Goal::Impl::add_transaction_replay_specs_to_goal(base::Transaction & transaction) {
+    libdnf5::GoalJobSettings settings = transaction_replay_settings;
+    bool skip_unavailable = settings.resolve_skip_unavailable(base->get_config());
+
+    for (const auto & package_replay : transaction_replay.packages) {
+        libdnf5::GoalJobSettings settings_per_package = settings;
+        settings_per_package.clean_requirements_on_remove = libdnf5::GoalSetting::SET_FALSE;
+        if (!package_replay.repo_id.empty()) {
+            repo::RepoQuery enabled_repos(base);
+            enabled_repos.filter_enabled(true);
+            enabled_repos.filter_id(package_replay.repo_id);
+            if (!enabled_repos.empty()) {
+                settings_per_package.to_repo_ids = {package_replay.repo_id};
+            }
+        }
+
+        if (package_replay.action == transaction::TransactionItemAction::UPGRADE ||
+            package_replay.action == transaction::TransactionItemAction::INSTALL ||
+            package_replay.action == transaction::TransactionItemAction::DOWNGRADE) {
+            if (package_replay.package_path.empty()) {
+                rpm_specs.emplace_back(GoalAction::INSTALL, package_replay.nevra, settings_per_package);
+            } else {
+                rpm_filepaths.emplace_back(GoalAction::INSTALL, package_replay.package_path, settings_per_package);
+            }
+            transaction.p_impl->rpm_reason_overrides[package_replay.nevra] = package_replay.reason;
+        } else if (package_replay.action == transaction::TransactionItemAction::REINSTALL) {
+            if (package_replay.package_path.empty()) {
+                rpm_specs.emplace_back(GoalAction::REINSTALL, package_replay.nevra, settings_per_package);
+            } else {
+                rpm_filepaths.emplace_back(GoalAction::REINSTALL, package_replay.package_path, settings_per_package);
+            }
+            transaction.p_impl->rpm_reason_overrides[package_replay.nevra] = package_replay.reason;
+        } else if (package_replay.action == transaction::TransactionItemAction::REMOVE) {
+            rpm_specs.emplace_back(GoalAction::REMOVE, package_replay.nevra, settings_per_package);
+            transaction.p_impl->rpm_reason_overrides[package_replay.nevra] = package_replay.reason;
+        } else if (package_replay.action == transaction::TransactionItemAction::REPLACED) {
+            if (!skip_unavailable) {
+                rpm_specs.emplace_back(GoalAction::REMOVE, package_replay.nevra, settings_per_package);
+            }
+        } else if (package_replay.action == transaction::TransactionItemAction::REASON_CHANGE) {
+            rpm_reason_change_specs.emplace_back(
```

Comment:
Shell we pass full NEVRA here or only name.arch of the package?

Reply:
It would make the `REASON_CHANGE` work even if a different version than expected is installed.
But why should it work differently than all other actions?

Reply:
This is a good question with no good answer.

I can argue that:
`dnf5 mark` with argument `dnf` will be impossible to replay when a different version is installed. We also know, that internal implementation does not care about version and only stores `name.arch` information.
In the following example the original version was not in any repository

```
$ sudo dnf5 mark dependency yum
$ dnf5 history store 61
$ sudo dnf5 upgrade yum
$ sudo dnf5 mark user  yum
$ dnf5 history replay transaction.json
Updating and loading repositories:
Repositories loaded.
Failed to resolve the transaction:
No match for argument: yum-0:4.18.1-20231110004827.0.g41a287e2.fc38.noarch
```

When I replay reason change (to dep) after I changed installed version and also after I changed reason (from dep to user) I got following problem.

```
dnf5 history replay transaction.json
Updating and loading repositories:
Repositories loaded.
Failed to resolve the transaction:
Argument 'yum-0:4.18.1-20231130005028.8.g6229c1d4.fc38.noarch' matches only excluded packages.
```

I think that the implementation in PR is more strict then it has to be. Then it is not robust to any changes of installed package when the reason change is replayed.